oauth2-server/src/main/java/com/clouway/oauth2/codechallenge/AuthorizationCodeVerifier.java

```diff
@@ -0,0 +1,53 @@
+package com.clouway.oauth2.codechallenge;
+
+import com.google.common.io.BaseEncoding;
+
+import java.nio.charset.StandardCharsets;
+import java.security.MessageDigest;
+import java.security.NoSuchAlgorithmException;
+
+/**
+ * @author Vasil Mitov <vasil.mitov@clouway.com>
+ */
+public class AuthorizationCodeVerifier implements CodeVerifier {
+  @Override
+  public boolean verify(CodeChallenge codeChallenge, String codeVerifier) {
+    try {
+      //no code verifier was provided and no code challenge was saved so this is a normal OAuth2 code flow without PKCE
+      if (isNullOrEmpty(codeVerifier) && isNullOrEmpty(codeChallenge)) {
+        return true;
+      }
+
+      if (isNullOrEmpty(codeVerifier) && !isNullOrEmpty(codeChallenge)) {
+        //there is a saved code challenge but provided codeVerifier is null or empty.
+        return false;
+      }
+
+      if (!isNullOrEmpty(codeVerifier) && isNullOrEmpty(codeChallenge)) {
+        //codeVerifier was provided but no codeChallenge was saved.
+        return false;
+      }
+
+      if (codeChallenge.method.equals("plain")) {
+        return codeChallenge.transformedCodeChallenge.equals(codeVerifier);
+      }
+
+      if (codeChallenge.method.equals("S256")) {
+        byte[] hashed = MessageDigest.getInstance("SHA-256").digest(codeVerifier.getBytes(StandardCharsets.UTF_8));
+        String transformedCodeVerifier = BaseEncoding.base64Url().encode(hashed);
+        return transformedCodeVerifier.equals(codeChallenge.transformedCodeChallenge);
+      }
+      return false;
+    } catch (NoSuchAlgorithmException e) {
+      return false;
+    }
+  }
+
+  private boolean isNullOrEmpty(String string) {
```

Comment:
guava has method isNullOrEmpty.

Reply:
Forgot about Strings.isNullOrEmpty. Implemented and removed method.